xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +132,760 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+						childPicker: b.childState.Picker,
+						// If both the `success_rate_ejection` and
+						// `failure_percentage_ejection` fields are unset in the
+						// configuration, the picker should not do that counting.
+						noopPicker: noopCfg,
+					},
+				})
+			case *LBConfig:
+				noopCfg := u.SuccessRateEjection == nil && u.FailurePercentageEjection == nil
+				if b.childState.Picker != nil && noopCfg != b.recentPickerNoop {
+					b.recentPickerNoop = noopCfg
+					b.cc.UpdateState(balancer.State{
+						ConnectivityState: b.childState.ConnectivityState,
+						// The outlier_detection LB policy will provide a picker that delegates to
+						// the child policy's picker, and when the request finishes, increment the
+						// corresponding counter in the map entry referenced by the subchannel
+						// wrapper that was picked.
+						Picker: &wrappedPicker{
+							childPicker: b.childState.Picker,
+							// If both the `success_rate_ejection` and
+							// `failure_percentage_ejection` fields are unset in the
+							// configuration, the picker should not do that counting.
+							noopPicker: noopCfg,
+						},
+					})
+				}
+			}
+		case <-b.closed.Done():
+			return
+		}
+	}
+}
+
+// numAddrsWithAtLeastRequestVolume returns the number of addresses present in
+// the map that have request volume of at least requestVolume.
+func (b *outlierDetectionBalancer) numAddrsWithAtLeastRequestVolume() uint32 {
+	var numAddrs uint32
+	for _, obj := range b.addrs {
+		if uint32(obj.callCounter.inactiveBucket.requestVolume) >= b.cfg.SuccessRateEjection.RequestVolume {
+			numAddrs++
+		}
+	}
+	return numAddrs
+}
+
+// meanAndStdDevOfSucceseesAtLeastRequestVolume returns the mean and std dev of
+// the number of requests of addresses that have at least requestVolume.
+func (b *outlierDetectionBalancer) meanAndStdDevOfSuccessesAtLeastRequestVolume() (float64, float64) {
+	// 2. Calculate the mean and standard deviation of the fractions of
+	// successful requests among addresses with total request volume of at least
+	// success_rate_ejection.request_volume.
+	var totalFractionOfSuccessfulRequests float64
+	var mean float64
+	for _, obj := range b.addrs {
+		// "of at least success_rate_ejection.request_volume"
+		if uint32(obj.callCounter.inactiveBucket.requestVolume) >= b.cfg.SuccessRateEjection.RequestVolume {
+			totalFractionOfSuccessfulRequests += float64(obj.callCounter.inactiveBucket.numSuccesses) / float64(obj.callCounter.inactiveBucket.requestVolume)
+		}
+	}
+	mean = totalFractionOfSuccessfulRequests / float64(len(b.addrs))
+	var sumOfSquares float64
+	for _, obj := range b.addrs {
+		devFromMean := (float64(obj.callCounter.inactiveBucket.numSuccesses) / float64(obj.callCounter.inactiveBucket.requestVolume)) - mean
+		sumOfSquares += devFromMean * devFromMean
+	}
+
+	variance := sumOfSquares / float64(len(b.addrs))
+	return mean, math.Sqrt(variance)
+
+}
+
+func (b *outlierDetectionBalancer) successRateAlgorithm() {
```

Comment:
Nit: `ejectLowSuccessRateBackends`? or something like that to indicate more precisely what the function is doing.  Same with the other `Algorithm` functions.